notifications_utils/recipients.py

```diff
@@ -58,6 +59,7 @@ def __init__(
         remaining_messages=sys.maxsize,
         allow_international_sms=False,
         allow_international_letters=False,
+        processing_timeout=10,
```

Comment:
Could you explain a bit more about the choice of processing_timeout as 10 seconds please? I can see from your graphs that almost all are uploaded within 10 seconds but equally I can see that some take longer than 10 seconds but less than 20. What is the argument for not upping this to 20 or some other number higher than 10? I guess we don't want it at 30 seconds because that is the http request timeout and so we probably need to put it somewhere a little under 30 seconds to account for any other activity in the http request that isn't spend running `recipients.rows`.

I acknowledge your commit message about how long do we think it is acceptable for a user to wait though so I am open to keeping it at 10. It might be good to get a second opinion on the default timeout just to make sure.

Reply:
My reasoning is just that: a website shouldn't take 30s to do something, or anything close to that.

Backing away from that extreme, slightly, I get that we want to give people a happy journey as much as possible, so let's look at the numbers for that graph. Worth noting some of these will be because of me testing.

- 0-10s: 99.86%
- 10-15s: 0.02%
- 15-20s: 0.01%
- 20-25s: 0.00%

Now we could argue: is there any harm in supporting an extra 0.03% of requests? Maybe not, but equally it's likely these spreadsheets / CSVs have something wrong with them, so showing an error a little sooner may actually be more helpful than making the user wait to see they can't have e.g. 10 million rows.

I am open to changing it. Given the tiny number of requests it would affect (some of which were probably mine), I'm also open to just trying it and seeing if it actually causes a problem for anyone.

@quis do you want to add an opinion here?

Reply:
Hmm, I guess it’s slightly arbitrary, in other words it’s hard to say why it should be 10 seconds (99.86%) or 15 seconds (99.88%).

My concern is that, let’s say 99.8% of sheets have a lot less than 100,000 rows. So hypothetically the difference between 10 and 15 seconds could be the difference between being able to process most 100,000 files and rejecting most of them. It wouldn’t be great if this change meant that we were unable to process most of the files that we say are within our row limit.

Could we do a quick look at how long it takes to process a file of 100,000 phone numbers (for example) on production, just to check we’re not getting close to timing out stuff we should be able to process?

I agree that having to wait 30 seconds is bad, but making people split their file in two so that we can process it is also an inconvenience.

Reply:
Another good point! I've done some tests in Production with 100K rows:

- Email CSV (1 column): average of 6.5s
- Email CSV (4 columns): average of 10s
- Phone CSV (4 columns): average of 13s
- Phone CSV (4 columns): average of 15s (with 32K errors for invalid numbers)
- Phone CSV (8 columns): average of 20s (with 32K errors)

I'm not sure how many columns we think is reasonable for a template. 4 seems reasonable to me, so this pushes us towards 20s, which means we may or may not timeout for the last scenario.

How do we feel about 20s?

Reply:
20 seconds feels good based on that.

It also matches what we tell people in tickets at the moment, eg ‘if your file has a lot of columns you may need to split it into multiple files’

Reply:
Cool, I've bumped it to 20s. The Admin app can also easily override it if it needs to, but it's handy to have a sensible default so we don't have to.

@idavidmcdonald that work for you?

Reply:
One more non-blocking comment: The API app will need to override this value. Otherwise you could have a situation where a CSV which the admin app happily processes in 19 seconds takes 21 seconds to process in the API app, causing the task to fail.

Reply:
I had a look into this when I was investigating this class. As far as I can tell, the `.rows` property isn't used in API, which [calls `.get_rows()` directly](https://github.com/alphagov/notifications-api/blob/19739945163793db79c89d89da39fc01c3c31b68/app/celery/tasks.py#L93). I'm not sure why it's written that way, but it does mean the change here should have no impact there.

I do realise that `.rows` is used by various other properties / methods in the class, but again I can't see anywhere in API where they're used - the class just isn't used that much in that repo.

Also agree it's non-blocking - I don't think we need to change the default in there though. I guess this does highlight that `processing_timeout` is a bit ambiguous, but I think that's more an issue with the class' implementation e.g. it _could_ be more like a factory that does all the processing upfront.

Short summary: I agree it's non-blocking, and don't think it's an issue.

Reply:
All good